functions/github/webhooks.js

```diff
@@ -0,0 +1,154 @@
+    if (eventType === "check_run") {
+      await handleCheckRunEvent(payload);
+    } else if (eventType === "pull_request") {
+      await handlePullRequestEvent(payload);
+    }
+  } catch (err) {
+    error("Failed to handle GitHub webhook",
+        {
+          error: err.message,
+          payload: payload,
+          eventType: eventType,
+        },
+    );
+    return res.status(500).send("Internal Server Error");
+  }
+
+  return res.status(200).send("OK");
+}
+
+/**
+  * Handles a GitHub pull request event.
+  *
+  * Updates the release state if the the event implies that the release state
+  * has changed, and needs to be synchronized. As of now, those two events are
+  * opens, and synchronizes (pushes to the pull request branch).
+  *
+  * @param {Object} payload The payload from the GitHub webhook.
+  * @return {Promise<void>}
+  */
+async function handlePullRequestEvent(payload) {
+  if (payload.action === "opened" || payload.action == "synchronize") {
+    const pullRequest = payload.pull_request;
+
+    if (pullRequest && pullRequest.head) {
+      const branchName = pullRequest.head.ref;
+
+      let releaseId;
+      try {
+        releaseId = await getReleaseIdFromBranch(branchName);
+      } catch (err) {
+        error("Error getting release ID from branch name",
+            {error: err.message});
+        return;
+      }
+
+      // If the releaseId exists, then a pull request was opened on a branch
+      // that is being tracked for a release.
+      //
+      // We assume that once a pull request is created on a release branch,
+      // the release configuration has been successfully generated.
+      // If this assumption is wrong, then the release will enter an error
+      // state.
+      if (releaseId) {
+        try {
+          log("Pull request updated on release branch",
+              {
+                releaseId: releaseId,
+                branchName: branchName,
+                pullRequest: pullRequest,
+              });
+          const octokit = new Octokit({auth: GITHUB_TOKEN.value()});
+          await syncReleaseState(releaseId, octokit);
+          log("Successfully synced release state", {releaseId: releaseId});
+        } catch (err) {
+          error("Failed to sync release state", {error: err.message});
+        }
+      }
+    }
+  }
+}
+
+
+/**
+  * Handles a GitHub check run event.
+  *
+  * @param {Object} payload The payload from the GitHub webhook.
+  * @return {Promise<void>}
+  */
+async function handleCheckRunEvent(payload) {
+  if (payload && payload.check_run) {
+    const checkRun = payload.check_run;
+
+    // If the check run is not for a release, then we can ignore it
+    if (checkRun.check_suite) {
+      const branchName = checkRun.check_suite.head_branch;
+
+      let releaseId;
+      try {
+        releaseId = getReleaseIdFromBranch(branchName);
+      } catch (err) {
+        error("Error getting release ID from branch name",
+            {error: err.message});
```

Comment:
Is there any instance in which this error message would change? Why not just error inside the function itself, instead of catching it at every call site?

Reply:
Yeah definitely, in this case, I think we could save some repetitive log statements by moving the logs to the function itself. But, I think this will contrast with the style of error handling I've tried to maintain for the codebase.

For all of the functions that interact with Firestore, I've let the errors propagate to the callers. This means that I usually only have to have one try/catch to handle the error, and be able to immediately return an appropriate HTTP response to clients (I don't do it in this case with webhooks, but it's useful for the other HTTP handlers.)

I believe this is the best approach because, for some reason, Firebase Functions return uncaught errors in HTTP responses automatically. So for example, if there's a mistake somewhere and an error is not caught when making a DB call when handling a request, the client will receive a response with the server-side Firestore error. So, having try/catches for everything in the handlers is more safe.